### ISSUE TITLE
Create models directory in AWQ with model classes

### ISSUE THREAD

**Issue by casperbh96** [mit-han-lab/llm-awq#71]
Ideally, when trying to support more models, you should create classes for each model to better manage them. This is an open issue to describe what should be moved into model classes in order to abstract the code a bit further and make it easier to support new models.

- Extract the [get_blocks](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/pre_quant.py#L22) function into model class
- Extract the [move_embed](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/pre_quant.py#L37) function into model class
- Extract the [scale_activations](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/quantizer.py#L14) function into model class
- Extract part of [auto_scale_block](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/auto_scale.py#L178) function into model class

**Comment by abhinavkulkarni** [comment]
I am echoing @casperbh96's points here as well.

[AutoGPTQ](https://github.com/PanQiWei/AutoGPTQ) may be a good repo to look at. First of all, there should be a `AutoModelForCausalLM.from_qunatized` method similar to `from_pretrained` method to load the AWQ models from a local directory / HF hub model repo. 

Secondly, such a repo should contain all the information that is needed to load the model - such as a `quantize_config.json` that AutoGPTQ maintains (for e.g. [here](https://huggingface.co/TheBloke/Llama-2-7b-Chat-GPTQ/blob/main/quantize_config.json)).

Thirdly, AWQ installation should be a one-command operation such as `pip install -e git@github.com/mit-han-lab/llm-awq.git` which should install CUDA kernels if needed or not (if only CPU is detected).

AWQ is slowly rising in popularity because a lot of people are finding it to be more accurate than GPTQ and faster. It would be good if these features are supported.

**Comment by casperbh96** [comment]
> [AutoGPTQ](https://github.com/PanQiWei/AutoGPTQ) may be a good repo to look at. First of all, there should be a `AutoModelForCausalLM.from_qunatized` method similar to `from_pretrained` method to load the AWQ models from a local directory / HF hub model repo. And corresponding `save_quantized` similar to `save_pretrained`.

Yes, the `from_pretrained`, `from_quantized`, `quantize`, and `save_quantized` methods would be amazing to have implemented for ease of use. I think this kind of functionality can come more easily after simply creating the directory of models that work.

> Secondly, such a repo should contain all the information that is needed to load the model - such as a `quantize_config.json` that AutoGPTQ maintains (for e.g. [here](https://huggingface.co/TheBloke/Llama-2-7b-Chat-GPTQ/blob/main/quantize_config.json)).

Yes, it would help to have the number of bits and group size available in a config.

> Thirdly, AWQ installation should be a one-command operation such as `pip install -e git@github.com/mit-han-lab/llm-awq.git` which should install CUDA kernels if needed or not (if only CPU is detected).

This is underway in my PR in #63 

> AWQ is slowly rising in popularity because a lot of people are finding it to be more accurate than GPTQ and faster. However, it is supported for fewer architectures (for e.g. no support for `GPTBigCodeForCausalLM` which startcoder and variants are based off). It would be good if these features are supported.

Yes, AWQ is in my experience the fastest and seemingly should be more accurate, although I still need more validation on this front.

**Comment by casperbh96** [comment]
@abhinavkulkarni I have created a draft PR #72 that I have gotten pretty far with. I will likely need some people to test it as the code is semi-close to being generally implemented. The last big step is re-implementing all the models, but should be easy assuming I did not make any larger mistakes while refactoring code.
